staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/construction.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/go-openapi/spec"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+// ResourceKind determines the scope of an API object: if it's the parent resource,
+// scale subresource or status subresource.
+type ResourceKind string
+
+const (
+	// Resource specifies an object of custom resource kind
+	Resource ResourceKind = "Resource"
+	// Scale specifies an object of custom resource's scale subresource kind
+	Scale ResourceKind = "Scale"
+	// Status specifies an object of custom resource's status subresource kind
+	Status ResourceKind = "Status"
+
+	scaleSchemaRef    = "#/definitions/io.k8s.api.autoscaling.v1.Scale"
+	statusSchemaRef   = "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.Status"
+	listMetaSchemaRef = "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.ListMeta"
+	patchSchemaRef    = "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.Patch"
+)
+
+// SwaggerConstructor takes in CRD OpenAPI schema and CustomResourceDefinitionSpec, and
+// constructs the OpenAPI swagger that an apiserver serves.
+type SwaggerConstructor struct {
+	// schema is the CRD's OpenAPI v2 schema
+	schema *spec.Schema
+
+	group    string
+	version  string
+	kind     string
+	listKind string
+	plural   string
+	scope    apiextensions.ResourceScope
+}
+
+// NewSwaggerConstructor creates a new SwaggerConstructor using the CRD OpenAPI schema
+// and CustomResourceDefinitionSpec
+func NewSwaggerConstructor(schema *spec.Schema, crdSpec *apiextensions.CustomResourceDefinitionSpec, version string) *SwaggerConstructor {
+	return &SwaggerConstructor{
+		schema:   schema,
+		group:    crdSpec.Group,
+		version:  version,
+		kind:     crdSpec.Names.Kind,
+		listKind: crdSpec.Names.ListKind,
+		plural:   crdSpec.Names.Plural,
+		scope:    crdSpec.Scope,
+	}
+}
+
+// ConstructCRDOpenAPISpec constructs the complete OpenAPI swagger (spec).
+func (c *SwaggerConstructor) ConstructCRDOpenAPISpec() *spec.Swagger {
```

Comment:
am a bit surprised this does not share anything with the native resource or kube-openapi.

Reply:
kube-openapi requires constructing a `*restful.WebService` to build the spec. We have a huge [function](https://github.com/kubernetes/kubernetes/blob/b1a52a38e9e3651680655416cc7afbec5e119854/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L170-L874) doing this for native resources. Now I look back and think maybe we can refactor it and reuse some code. I will give it a try as a follow-up.

Efficiency of filling a template might be another minor reason

Reply:
done